### PR TITLE
Allow halos to bleed over periodic BCs

### PR DESCRIPTION
**Pull request by permcody**
This is another non-trivial update to the GrainTracker that allows halos to bleed over a periodic BC. This works by employing some of the "topological neighbor" methods in libMesh for getting halos. However, it's tricky to do this correctly with mesh adaptivity and recursion. I'm maintaining a separate list of IDs that are "disjoint" from the regular halos. I have to run a "connected component" algorithm on those halos to put them into separate buckets, then build new bounding boxes around those regions.

Including a new test and images showing the difference.

### CODE REVIEW

**Comment by permcody** [comment]
grains_near_edge

BEFORE:
halos_no_bleed

AFTER:
halos_bleed_across_bc

**Comment by permcody** [comment]
With @dschwen likely out for a little while, I may need to have @jwpeterson review this and merge. While I don't necessarily need this for the _next_ PR I have lined up (to overhaul polycrystal ICs). My two PRs are changing logic in several of the same areas so if this PR requires changes on its own, it'll be a rebasing challenge.

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 1]
```diff
@@ -1521,5 +1660,45 @@ operator<<(std::ostream & out, const FeatureFloodCount::FeatureData & feature)
...
+
+void
+updateBBoxExtremesHelper(MeshTools::BoundingBox & bbox, const Elem & elem)
+{
+  for (auto node_n = decltype(elem.n_nodes())(0); node_n < elem.n_nodes(); ++node_n)
+    updateBBoxExtremesHelper(bbox, *(elem.get_node(node_n)));
+}
+
+bool
+areElemListsMergeable(std::list<dof_id_type> elem_list1,
```
Comment (jwpeterson) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Do you really want to pass these lists to this function by value? Seems like they could be (const?) references...

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h [hunk 2]
```diff
@@ -185,6 +184,17 @@ class FeatureFloodCount : public GeneralPostprocessor,
      */
     bool boundingBoxesIntersect(const FeatureData & rhs) const;
 
+    /**
+     * The routine called to see if two features are mergable:
```
Comment (jwpeterson) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: I don't know how "mergeable" is actually spelled, but the documentation spelling should at least agree with the code spelling.

**Review by jwpeterson — CHANGES_REQUESTED** [review]
I do like how that one crazy ass if-statement is gone now...

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 1 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Definitely don't want to make copies!

**Review by jwpeterson — APPROVED** [review]

**Comment by brianmoose** [comment]
Looks like this causes the `phase_field/examples/grain_growth.grain_growth_2D_graintracker` test to fail on mac (in opt mode). Seems to be fine on linux.
https://www.moosebuild.org/job/82567/

**Comment by permcody** [comment]
Yes I am aware. The issue that "check_syntax" does a lot more than parse syntax. So when a test was put in there to check the syntax and change the problem to make it a little smaller, this issue was triggered. I have another PR that will take care of this once I shake out the failures it caused #8948.

**Pull request merged**